add updated bellatrix/capella vectors

njgheorghita opened the pull request:
New test vectors for `HeaderWithProof` content key/values using proofs from `./tests/mainnet/history/headers_with_proof/block_proofs_*`, with the updated `HeaderWithProof` types. Also includes the recently removed `Union` change from the `HWP` ssz encoding. 

If other teams could confirm/deny that these are valid vectors, that would be great!

fearlessfe commented:
I have some problem when decode capella content value. Decode content value to `BlockHeaderWithProof ` is ok, when I trying to decode `BlockHeaderWithProof .proof` to `BlockProofHistoricalSummaries `, it has error.
```
# Proof that EL block_hash is in BeaconBlock -> BeaconBlockBody -> ExecutionPayload
ExecutionBlockProofCapella = List[Bytes32, limit=12]

# Proof that BeaconBlock root is part of historical_summaries and thus canonical
# For Capella and onwards
BeaconBlockProofHistoricalSummaries = Vector[Bytes32, 13]

# Proof for EL BlockHeader for Capella and onwards
BlockProofHistoricalSummaries = Container[
    beaconBlockProof: BeaconBlockProofHistoricalSummaries, # Proof that the BeaconBlock is part of the historical_summaries and thus part of the canonical chain
    beaconBlockRoot: Bytes32, # hash_tree_root of BeaconBlock used to verify the proofs
    executionBlockProof: ExecutionBlockProofCapella, # Proof that EL BlockHash is part of the BeaconBlock
    slot: Slot # Slot of BeaconBlock, used to calculate the historical_summaries index
]

BlockHeaderWithProof = Container(
  header: ByteList[MAX_HEADER_LENGTH], # RLP encoded header in SSZ ByteList
  proof: ByteList[MAX_HEADER_PROOF_LENGTH], # SSZ encoded proof
)
```
The error occured when decode field `BlockProofHistoricalSummaries .executionBlockProof`, it will read a 4 bytes offset, but the value is bigger than the proof size, so decode failed.

morph-dev commented:
@fearlessfe, can you share code that you are using? How do you define `ExecutionBlockProofCapella` type?

It's worth highlighting that this type (unlike most others) is not ssz Vector, but ssz List (with upper limit of 12 elements, not fixed at 12 elements).

morph-dev commented:
Or even better, create the test vectors yourself and share so we can compare.
I believe @njgheorghita used blocks and data from here: https://github.com/ethereum/portal-spec-tests/tree/master/tests/mainnet/history/headers_with_proof/block_proofs_capella, so you should be able to do the same.

fearlessfe commented:
we use golang, the struct define are as below, and  `UnmarshalSSZ` is generate by fastssz lib. 

```
type BlockProofHistoricalSummaries struct {
	BeaconBlockProof    [][]byte `ssz-size:"13,32"`
	BeaconBlockRoot     []byte   `ssz-size:"32"`
	ExecutionBlockProof [][]byte `ssz-size:"?,32" ssz-max:"12"`
	Slot                uint64
}

func (b *BlockProofHistoricalSummaries) UnmarshalSSZ(buf []byte) error {
	var err error
	size := uint64(len(buf))
	if size < 460 {
		return ssz.ErrSize
	}

	tail := buf
	var o2 uint64

	// Field (0) 'BeaconBlockProof'
	b.BeaconBlockProof = make([][]byte, 13)
	for ii := 0; ii < 13; ii++ {
		if cap(b.BeaconBlockProof[ii]) == 0 {
			b.BeaconBlockProof[ii] = make([]byte, 0, len(buf[0:416][ii*32:(ii+1)*32]))
		}
		b.BeaconBlockProof[ii] = append(b.BeaconBlockProof[ii], buf[0:416][ii*32:(ii+1)*32]...)
	}

	// Field (1) 'BeaconBlockRoot'
	if cap(b.BeaconBlockRoot) == 0 {
		b.BeaconBlockRoot = make([]byte, 0, len(buf[416:448]))
	}
	b.BeaconBlockRoot = append(b.BeaconBlockRoot, buf[416:448]...)

	// Offset (2) 'ExecutionBlockProof'
	if o2 = ssz.ReadOffset(buf[448:452]); o2 > size {
		return ssz.ErrOffset
	}

	if o2 != 460 {
		return ssz.ErrInvalidVariableOffset
	}

	// Field (3) 'Slot'
	b.Slot = ssz.UnmarshallUint64(buf[452:460])

	// Field (2) 'ExecutionBlockProof'
	{
		buf = tail[o2:]
		num, err := ssz.DivideInt2(len(buf), 32, 12)
		if err != nil {
			return err
		}
		b.ExecutionBlockProof = make([][]byte, num)
		for ii := 0; ii < num; ii++ {
			if cap(b.ExecutionBlockProof[ii]) == 0 {
				b.ExecutionBlockProof[ii] = make([]byte, 0, len(buf[ii*32:(ii+1)*32]))
			}
			b.ExecutionBlockProof[ii] = append(b.ExecutionBlockProof[ii], buf[ii*32:(ii+1)*32]...)
		}
	}
	return err
}
```

fearlessfe commented:
@morph-dev I know the difference, the field sequence is difference.in the origin [17034870 block file](https://github.com/ethereum/portal-spec-tests/blob/master/tests/mainnet/history/headers_with_proof/block_proofs_capella/beacon_block_proof-17034870.yaml).  the `BlockProofHistoricalSummaries. beaconBlockProof` should be the `historical_summaries_proof` field in yaml, should be the first field in `BlockProofHistoricalSummaries`, but the content of `historical_summaries_proof` is appeared last in the content value

njgheorghita commented:
Yup, I think I got these mixed up. Will update the vectors now.

njgheorghita commented:
@fearlessfe Thanks for helping out, sorry for the confusion. I've updated the vectors, if you could see if the new versions are compatible with your tooling? I also updated the keys in the yaml test vector files to match the specs more closely

fearlessfe commented:
@njgheorghita  The new vectors works fine

kdeme reviewed (approved):
test vectors work for fluffy
